lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,923 @@
+        :param versions: A list of versions of the collection that is available
+        :param requirement: The version requirement string used to verify the list of versions fit the requirements.
+        :param force: Whether the force flag applied to the collection.
+        :param parent: The name of the parent the collection is a dependency of.
+        :param validate_certs: Whether to validate the Galaxy server certificate.
+        :param metadata: The collection metadata dict if it has already been retrieved.
+        :param files: The files that exist inside the collection. This is based on the FILES.json file inside the
+            collection artifact.
+        :param skip: Whether to skip installing the collection, should be set if the collection is already installed
+            and force is not set.
+        """
+        self.namespace = namespace
+        self.name = name
+        self.path = path
+        self.source = source
+        self.versions = set(versions)
+        self.force = force
+        self.skip = skip
+        self.required_by = []
+        self._validate_certs = validate_certs
+
+        self._metadata = metadata
+        self._files = files
+        self._galaxy_info = None
+
+        self.add_requirement(parent, requirement)
+
+    def __str__(self):
+        return "%s.%s" % (self.namespace, self.name)
+
+    @property
+    def latest_version(self):
+        versions = [v for v in self.versions if v != '*']
+        if versions:
+            versions = sorted(versions, key=LooseVersion)
+            return versions[-1]
+        else:
+            return '*'
+
+    @property
+    def dependencies(self):
+        if self._metadata:
+            return self._metadata['dependencies']
+        elif len(self.versions) > 1:
+            return None
+
+        self._get_metadata()
+        return self._metadata['dependencies']
+
+    def add_requirement(self, parent, requirement):
+        self.required_by.append((parent, requirement))
+        new_versions = set([v for v in self.versions if self._meets_requirements(v, requirement)])
+        if len(new_versions) == 0:
+            if self.skip:
+                force_flag = '--force-with-deps' if parent else '--force'
+                msg = "Cannot meet requirement %s:%s as it is already installed. Use %s to overwrite" \
+                      % (str(self), requirement, force_flag)
+                raise AnsibleError(msg)
+            elif parent is None:
+                msg = "Cannot meet requirement %s for dependency %s" % (requirement, str(self))
+            else:
+                msg = "Cannot meet dependency requirement '%s:%s' for collection %s" % (str(self), requirement, parent)
+
+            collection_source = self.path or self.source
+            req_by = "\n".join(
+                ["\t%s - '%s:%s'" % (to_text(p) if p else 'base', str(self), r) for p, r in self.required_by]
```

Comment:
str.join() accepts an iterable. You use a list comprehension which first creates a `list()` in memory and then gets fed to `str.join()`. You can use a generator expression which will gen unwinded by `str.join()` internally without consuming memory for an intermediate list representation.

```suggestion
                "\t%s - '%s:%s'" % (to_text(p) if p else 'base', str(self), r)
                for p, r in self.required_by
```